lib/sqlalchemy/testing/provision.py

```diff
@@ -98,214 +95,43 @@ def _configs_for_db_operation():
 
 
 @register.init
-def _create_db(cfg, eng, ident):
+def create_db(cfg, eng, ident):
+    """write a docstring """
     raise NotImplementedError("no DB creation routine for cfg: %s" % eng.url)
 
```

Comment:
**mike bayer** (zzzeek) wrote:

can we put a quick docstring for each of these?

Reply:
I can give it a go. Am I right in thinking that a "follower" is one of the databases that is created on the fly when tests are run using multiple processes? The `dburi` for the test run specifies the "leader" database and the "followers" are created dynamically based on certain characteristics of the "leader"?

Reply:
yes,  that is what it means.     

the docstrings here could just express what the name of the function says for now, nothing fancy just something presentable :)

Reply:
How do these look?

```diff
Index: lib/sqlalchemy/testing/provision.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lib/sqlalchemy/testing/provision.py	(revision 01436b325db0031f149c5cd58983b49bc12a03d0)
+++ lib/sqlalchemy/testing/provision.py	(date 1579908754539)
@@ -98,37 +98,50 @@
 
 @register.init
 def create_db(cfg, eng, ident):
-    """write a docstring """
+    """Dynamically create a database for testing.
+
+    Used when a test run will employ multiple processes, e.g., when run
+    via `tox` or `py.test -n4`.
+    """
     raise NotImplementedError("no DB creation routine for cfg: %s" % eng.url)
 
 
 @register.init
 def drop_db(cfg, eng, ident):
-    """write a docstring """
+    """Drop a database that we created for testing."""
     raise NotImplementedError("no DB drop routine for cfg: %s" % eng.url)
 
 
 @register.init
 def update_db_opts(db_url, db_opts):
-    """write a docstring """
+    """Set database options (db_opts) for a test database that we created.
+    """
     pass
 
 
 @register.init
 def configure_follower(cfg, ident):
-    """write a docstring """
+    """Dialect-specific configuration settings for a follower database."""
     pass
 
 
 @register.init
 def post_configure_engine(url, engine, follower_ident):
-    """write a docstring """
+    """Optional extra steps after configuring an engine for testing.
+
+    (For the internal dialects, currently only used by sqlite.)
+    """
     pass
 
 
 @register.init
 def follower_url_from_main(url, ident):
-    """write a docstring """
+    """Create a connection URL for a dynamically-created test database.
+
+    Arguments:
+    url -- the connection URL specified when the test run was invoked
+    ident -- the dynamically-generated name of the database to be created
+    """
     url = sa_url.make_url(url)
     url.database = ident
     return url
@@ -164,7 +177,9 @@
 
 @register.init
 def temp_table_keyword_args(cfg, eng):
-    """dialect-specific implementations of this method will return the
+    """Specify keyword arguments for creating a temporary Table.
+
+    Dialect-specific implementations of this method will return the
     kwargs that are passed to the Table method when creating a temporary
     table for testing, e.g., in the define_temp_tables method of the
     ComponentReflectionTest class in suite/test_reflection.py
```

Reply:
p.s. When the time comes to commit I think I'll move `configure_follower` below `follower_url_from_main` to improve readability.

Reply:
it's a great start.  Use `:param url:` format for those parameters.   also "ident" is likely best referred to as a "worker identifier" since that's what pytest-xdist calls it.

Reply:
**Gord Thompson** (gordthompson) wrote:

Done